alf/utils/process_coordinator_test.py

```diff
@@ -0,0 +1,432 @@
+        self.assertTrue(reported)
+        # If we clear_stop the exceptions are handled normally.
+        coord.clear_stop()
+        try:
+            raise RuntimeError("After clear")
+        except RuntimeError:
+            coord.request_stop(sys.exc_info())
+        with self.assertRaisesRegex(RuntimeError, "After clear"):
+            coord.join([])
+
+
+def _stop_at_0(coord, n, m=None):
+    if n.value == 0:
+        coord.request_stop()
+    else:
+        if m:
+            m.decrement()
+        n.value -= 1
+
+
+class ProcessTest(test.TestCase):
+    def testTargetArgs(self):
+        n = Value(ctypes.c_int, 3)
+        coord = coordinator.Coordinator()
+        p = coordinator.Process.process(
+            coord, target=_stop_at_0, args=(coord, n))
+        coord.join([p])
+        self.assertEqual(0, n.value)
+
+    def testTargetKwargs(self):
+        n = Value(ctypes.c_int, 3)
+        coord = coordinator.Coordinator()
+        p = coordinator.Process.process(
+            coord, target=_stop_at_0, kwargs={
+                "coord": coord,
+                "n": n
+            })
+        coord.join([p])
+        self.assertEqual(0, n.value)
+
+    def testTargetMixedArgs(self):
+        n = Value(ctypes.c_int, 3)
+        coord = coordinator.Coordinator()
+        p = coordinator.Process.process(
+            coord, target=_stop_at_0, args=(coord, ), kwargs={"n": n})
+        coord.join([p])
+        self.assertEqual(0, n.value)
+
+    def testInheritedTarget(self):
+        class MyProcess(coordinator.Process):
+            def __init__(self, coord, args=(), kwargs={}):
+                super().__init__(coord, args=args, kwargs=kwargs)
+
+            def body(self, n=None):
+                _stop_at_0(self._coord, n)
+
+        n = Value(ctypes.c_int, 3)
+        coord = coordinator.Coordinator()
+        p = MyProcess(coord, kwargs={"n": n})
+        p.start()
+        coord.join([p])
+        self.assertEqual(0, n.value)
+
+    def testModelSharing(self):
+        class MyProcess(coordinator.Process):
+            def __init__(self, coord, args=(), kwargs={}):
+                super().__init__(coord, args=args, kwargs=kwargs)
+                self.n = Value(ctypes.c_int, 3)
+
+            def body(self, m=None):
+                _stop_at_0(self._coord, self.n, m)
+
+        from alf.algorithms.algorithm import Algorithm
+
+        class MyAlgorithm(Algorithm):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer('_m', torch.tensor(0, dtype=torch.int32))
+                self.x = torch.tensor(0, dtype=torch.int32)
+
+            def decrement(self):
+                self._m -= 1
+                self.x -= 1
+
+        m = MyAlgorithm()
+        m.share_memory()
+        coord = coordinator.Coordinator()
+        p = MyProcess(coord, kwargs={"m": m})
+        # A change in parent process is reflected in child process
+        # via share_memory
+        m._m.fill_(-1)
```

Comment:
no need to have this line; just check m._m equal to -3 after join()

Reply:
This is to test after subprocess p is already created, when we change model _m on parent process, the change is actually reflected in the subprocess.

None of the existing tests do that test.

Reply:
If you want to do that, then this line should be moved after the process starts. Right now, the child process will inherit this -1.